bitcoin/src/network/mod.rs

```diff
@@ -57,27 +57,75 @@ impl From<Network> for NetworkKind {
 
         match n {
             Bitcoin => NetworkKind::Main,
-            Testnet | Signet | Regtest => NetworkKind::Test,
+            Testnet(_) | Signet | Regtest => NetworkKind::Test,
         }
     }
 }
 
-/// The cryptocurrency network to act on.
+/// The testnet version to act on.
 #[derive(Copy, PartialEq, Eq, PartialOrd, Ord, Clone, Hash, Debug)]
 #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
 #[cfg_attr(feature = "serde", serde(rename_all = "lowercase"))]
 #[non_exhaustive]
+pub enum TestnetVersion {
+    /// Testnet version 3.
+    V3,
+    /// Testnet version 4.
+    /// This is the latest testnet version.
+    V4,
+}
+
+/// The cryptocurrency network to act on.
+#[derive(Copy, PartialEq, Eq, PartialOrd, Ord, Clone, Hash, Debug)]
+#[non_exhaustive]
 pub enum Network {
     /// Mainnet Bitcoin.
     Bitcoin,
     /// Bitcoin's testnet network.
-    Testnet,
+    Testnet(TestnetVersion),
     /// Bitcoin's signet network.
     Signet,
     /// Bitcoin's regtest network.
     Regtest,
 }
 
+#[cfg(feature = "serde")]
+impl Serialize for Network {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(self.as_display_str())
+    }
+}
+
+#[cfg(feature = "serde")]
+impl<'de> Deserialize<'de> for Network {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct NetworkVisitor;
+
+        impl<'de> Visitor<'de> for NetworkVisitor {
+            type Value = Network;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("a valid network identifier")
+            }
+
+            fn visit_str<E>(self, value: &str) -> Result<Network, E>
+            where
+                E: serde::de::Error,
+            {
+                Network::from_str_impl(value).map_err(E::custom)
```

Comment:
In ae178a392f9774f31195aff3b49830e9c00fb3e7:

Why introduce the `from_str_impl` method? Why not just call `from_str` directly?

Reply:
It'd make sense if the method accepted `s: impl AsRef<str> + Into<String>` so that `to_owned` is converted to `into` and then also have `visit_string` to save allocation in error case.

Reply:
Thank you very much for your valuable advice. It has been adjusted.